diff --git a/kernel/module/strict_rwx.c b/kernel/module/strict_rwx.c
index 397e18f..575bf99 100644
--- a/kernel/module/strict_rwx.c
+++ b/kernel/module/strict_rwx.c
@@ -46,6 +46,7 @@ void module_disable_ro(const struct module *mod)
 	module_set_memory(mod, MOD_INIT_TEXT, set_memory_rw);
 	module_set_memory(mod, MOD_RODATA, set_memory_rw);
 	module_set_memory(mod, MOD_INIT_RODATA, set_memory_rw);
+	module_set_memory(mod, MOD_RO_AFTER_INIT, set_memory_rw);
 }
 #endif /* CONFIG_LIVEPATCH_WO_FTRACE */
 
